### PR TITLE
Add typed variant of `ExtensionContainer.findByName` function

### LINKED ISSUE #17254: Add typed variant of `ExtensionContainer.findByName` function
It would be great to have a typed variant of `ExtensionContainer.findByName` function.
I've created this extension-function for my own project:

```kotlin
@Suppress("extension_shadowed_by_member")
inline fun <reified T : Any> ExtensionContainer.findByName(name: String): T? =
    findByName(name)?.let {
        it as? T
            ?: error(
                "Element '$name' of type '${it::class.java.name}' from container '$this' cannot be cast to '${T::class.qualifiedName}'.",
            )
    }
```

If this extension-function is correct, I can create Pull Request with this.

### Expected Behavior

There should be a typed variant of `findByName` similar to `getByName`:
https://github.com/gradle/gradle/blob/06140ac5fbdce282f868197064156b9b1753012f/subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/ExtensionContainerExtensions.kt#L50

### Current Behavior

There no such function.

### Context

- Gradle 7.0.2
- Gradle Kotlin DSL

### PR DESCRIPTION
**Pull request by osipxd**
Fixes #17254

This pull request adds typed variant of function `ExtensionContainer.findByName`.
It can be used in the following way:

```kotlin
val myExtension = extensions.findByName<MyExtension>("my") ?: extension.create<MyExtension>("my")
```

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [X] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew kotlin-dsl:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by osipxd** [comment]
I'm not sure if I should write tests for this function because `getByName` is not covered with tests. Also `ExtensionContaierExtensionsTest` contains only one test for each function and these tests not cover corner cases.

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/ExtensionContainerExtensions.kt [hunk 1]
```diff
@@ -42,18 +43,40 @@ operator fun ExtensionContainer.get(name: String): Any =
...
+inline fun <reified T : Any> ExtensionContainer.findByName(name: String): T? =
+    findByName(name)?.let { tryCastExtension(it, name) }
+
+
+/** Casts given [extension] to the specified type [T] or throws [IllegalStateException]. */
+@PublishedApi
+internal
+inline fun <reified T : Any> ExtensionContainer.tryCastExtension(extension: Any, name: String): T =
+    extension as? T
+        ?: throw IllegalStateException(
```
Comment (osipxd) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/ExtensionContainerExtensions.kt:77]: Can I use here function `error("...")` instead of `throw IllegalStateException`? Not seen this function in the project source code.

**Comment by osipxd** [comment]
@bmuskalla, can you take a look at this?

**Comment by bmuskalla** [comment]
Maybe @ljacomet ?

**Comment by osipxd** [comment]
Rebased to the master with the hope this PR will be merged someday

**Comment by osipxd** [comment]
🙃 Should I rebase this?

**Comment by eskatos** [comment]
Hi @osipxd,

Thank you for the PR. Sorry it took us a ridiculous amount of time to get back to you.

We will not merge this and prefer the solution in this issue:
* https://github.com/gradle/gradle/issues/25342

Closing this PR

**Comment by osipxd** [comment]
Oh, glad  to hear the issue will be fixed :) Thank you
